testsuites/CBLTester/CBL_Functional_tests/TestSetup_FunctionalTests/test_replication_eventing.py

```diff
@@ -0,0 +1,511 @@
+    channels = ["ABC"]
+    username = "autotest"
+    password = "password"
+
+    # Create CBL database
+    sg_client = MobileRestClient()
+
+    # Reset cluster to ensure no data in system
+    c = cluster.Cluster(config=cluster_config)
+    c.reset(sg_config_path=sg_config)
+
+    # 1.1 Creating Docs in CBL
+    cbl_doc_ids = db.create_bulk_docs(num_of_docs, "push_cbl_docs", db=cbl_db, channels=channels)
+
+    # 1.2 Creating Docs in SG
+    sg_client.create_user(url=sg_admin_url, db=sg_db, name=username, password=password, channels=channels)
+    cookie, session = sg_client.create_session(url=sg_admin_url, db=sg_db, name=username)
+    auth_session = cookie, session
+    sync_cookie = "{}={}".format(cookie, session)
+    session_header = {"Cookie": sync_cookie}
+
+    sg_docs = sg_client.add_docs(url=sg_url, db=sg_db, number=num_of_docs, id_prefix="pull_sg_docs", channels=channels,
+                                 auth=auth_session)
+    sg_doc_ids = [doc["id"] for doc in sg_docs]
+
+    # 2 Adding Listener for replicator
+    replicator = Replication(base_url)
+    repl_config = replicator.configure(source_db=cbl_db,
+                                       target_url=sg_blip_url,
+                                       continuous=False,
+                                       headers=session_header)
+    repl = replicator.create(repl_config)
+    repl_change_listener = replicator.addReplicatorEventChangeListener(repl)
+
+    # 3. Starting Replication and waiting for it finish
+    replicator.start(repl)
+    replicator.wait_until_replicator_idle(repl, max_times=5000)
+
+    # 4. Getting changes from the replication event listener
+    doc_repl_event_changes = replicator.getReplicatorEventChanges(repl_change_listener)
+    replicator.removeReplicatorEventListener(repl, repl_change_listener)
+    replicator.stop(repl)
+
+    # Processing received events
+    replicated_event_changes = get_event_changes(doc_repl_event_changes)
+    push_docs = []
+    pull_docs = []
+    error_docs = []
+    for doc in replicated_event_changes:
+        if replicated_event_changes[doc]['push'] is True:
+            push_docs.append(doc)
+        else:
+            pull_docs.append(doc)
+        if replicated_event_changes[doc]['error_code'] is not None:
+            error_docs.append({doc: "{}, {}".format(replicated_event_changes[doc]['error_code'],
+                                                    replicated_event_changes[doc]['error_domain'])})
+
+    # 5. Validating the event counts and verifying the push and pull event against doc_ids
+    assert sorted(push_docs) == sorted(
+        cbl_doc_ids), "Replication event push docs are not equal to expected no. of docs to be pushed"
+    assert sorted(pull_docs) == sorted(
+        sg_doc_ids), "Replication event pull docs are not equal to expected no. of docs to be pulled"
+    assert len(error_docs) == 0, "Error found in replication events {}".format(error_docs)
+
+
+@pytest.mark.sanity
```

Comment:
remove sanity here. we need sanity only for one positive flow test

Reply:
removed